c/cardano_lock.c

```diff
@@ -1,18 +1,346 @@
+
+  return nanocbor_encoded_len(&enc);
+}
+
+static int _make_cursor(size_t index, size_t source, size_t len,
+                        mol2_source_t read, mol2_cursor_t *cur) {
+  ASSERT(cur);
+  ASSERT(len);
+  ASSERT(read);
+
+  int err = 0;
+
+  cur->offset = 0;
+  cur->size = len;
+
+  memset(g_mol_data_source, 0, sizeof(g_mol_data_source));
+  mol2_data_source_t *ptr = (mol2_data_source_t *)g_mol_data_source;
+
+  ptr->read = read;
+  ptr->total_size = len;
+
+  ptr->args[0] = index;
+  ptr->args[1] = source;
+
+  ptr->cache_size = 0;
+  ptr->start_point = 0;
+  ptr->max_cache_size = MAX_CACHE_SIZE;
+  cur->data_source = ptr;
+
+  return err;
+}
+
+static uint32_t _read_data_from_witness(uintptr_t arg[], uint8_t *ptr,
+                                        uint32_t len, uint32_t offset) {
+  int err;
+  uint64_t output_len = len;
+  err = ckb_load_witness(ptr, &output_len, offset, arg[0], arg[1]);
+  if (err != 0) {
+    return 0;
+  }
+  if (output_len > len) {
+    return len;
+  } else {
+    return (uint32_t)output_len;
+  }
+}
+
+int _get_cursor_from_witness(WitnessArgsType *witness, size_t index,
+                             size_t source) {
+  int err = CKB_SUCCESS;
+  mol2_cursor_t cur;
+  uint64_t len = 0;
+  ckb_load_witness(NULL, &len, 0, index, source);
+  if (len == 0) {
+    return ERROR_LOAD_WITNESS;
+  }
+
+  err = _make_cursor(index, source, len, _read_data_from_witness, &cur);
+  if (err != CKB_SUCCESS) {
+    return err;
+  }
+
+  *witness = make_WitnessArgs(&cur);
+  return err;
+}
 
 #ifdef CKB_USE_SIM
 int simulator_main() {
 #else
-int main(int argc, const char* argv[]) {
+int main(int argc, const char *argv[]) {
 #endif
-  return 0;
-}
+  // Get public key
+  uint8_t msg[32] = {0};
+  int error = generate_sighash_all(msg, sizeof(msg));
+  if (error != CKB_SUCCESS) {
+    return error;
+  }
+
+  size_t new_msg_len = generate_new_msg(NULL, 0, msg, sizeof(msg), NULL, 0);
+  if (new_msg_len == 0) {
+    return ERROR_GENERATE_NEW_MSG;
+  }
+  uint8_t new_msg[new_msg_len];
+  memset(new_msg, 0, new_msg_len);
+  if (generate_new_msg(new_msg, new_msg_len, msg, sizeof(msg), NULL, 0)) {
+    return ERROR_GENERATE_NEW_MSG;
+  }
+
+  // get lock args:
+  uint8_t identity[21] = {0};
+
+  unsigned char script[SCRIPT_SIZE];
+  uint64_t scritp_len = SCRIPT_SIZE;
+  error = ckb_load_script(script, &scritp_len, 0);
+  if (error != CKB_SUCCESS || scritp_len > SCRIPT_SIZE) {
+    return ERROR_LOAD_SCRIPT;
+  }
+  mol_seg_t script_seg;
+  script_seg.ptr = (uint8_t *)script;
+  script_seg.size = scritp_len;
 
+  mol_seg_t args = MolReader_Script_get_args(&script_seg);
+  mol_seg_t args_raw_bytes = MolReader_Bytes_raw_bytes(&args);
+  if (args_raw_bytes.size < sizeof(identity)) {
+    return ERROR_LOAD_SCRIPT;
+  }
+  memcpy(identity, args_raw_bytes.ptr, sizeof(identity));
 
+  // uint8_t witness_data[96] = {0};
+  // uint64_t witness_data_len = sizeof(witness_data);
+  for (size_t index = 0;; index++) {
```

Comment:
Don't need to loop through all index: Just only one witness is used. see: https://github.com/nervosnetwork/ckb-system-scripts/blob/e08e6016f16072fc2f44cf889ae063fa5b7e10c7/c/secp256k1_blake160_sighash_all.c#L110-L120